src/eckit/linalg/allocator/InPlaceAllocator.cc

```diff
@@ -0,0 +1,57 @@
+/*
+ * (C) Copyright 2024- ECMWF.
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ * In applying this licence, ECMWF does not waive the privileges and immunities
+ * granted to it by virtue of its status as an intergovernmental organisation
+ * nor does it submit to any jurisdiction.
+ */
+
+
+#include "eckit/linalg/allocator/InPlaceAllocator.h"
+
+#include <ostream>
+
+#include "eckit/exception/Exceptions.h"
+
+
+namespace eckit::linalg::allocator {
+
+
+InPlaceAllocator::InPlaceAllocator(Size Nr, Size Nc, Size nnz, Index* ia, Index* ja, Scalar* a) :
+    Nr_(Nr), Nc_(Nc), nnz_(nnz), ia_(ia), ja_(ja), a_(a) {
+    ASSERT(ia_ != nullptr);
+    ASSERT(ja_ != nullptr);
+    ASSERT(a_ != nullptr);
+}
+
+
+SparseMatrix::Layout InPlaceAllocator::InPlaceAllocator::allocate(SparseMatrix::Shape& shape) {
+    shape.size_ = nnz_;
+    shape.rows_ = Nr_;
+    shape.cols_ = Nc_;
+
+    SparseMatrix::Layout layout;
+    layout.outer_ = reinterpret_cast<decltype(SparseMatrix::Layout::outer_)>(ia_);
```

Comment:
I believe a reinterpret_cast or const_cast or any other casting should not be necessary.

Reply:
I understood now that the reinterpret_cast is a byproduct of a PR from a few months back where the data type of outer_ has changed from Index to UIndex.